TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+{
+    NSNumber *number = schema[key];
+    if (!number) {
+        return nil;
+    }
+
+    [self pushPathComponent:key];
+    [self checkObject:number isOfClass:[NSNumber class] allowNil:YES];
+    if ([number doubleValue] <= 0) {
+        NSString *reason = [NSString stringWithFormat:@"Invalid value for key %@. Expected number greater than zero.", key];
+        [self failWithErrorCode:TWTJSONSchemaParserErrorCodeInvalidValue reason:reason object:number];
+    }
+
+    [self popPathComponent];
+    return number;
+}
+
+
+// Keywords: min & maxLength, min & maxItems, min & maxProperties
+- (NSNumber *)parseUnsignedIntegerFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSNumber *number = schema[key];
+    [self checkObject:number isOfClass:[NSNumber class] allowNil:YES];
+
+    if (number) {
+        double numberValue = [number doubleValue];
+        BOOL isUnsignedInteger = numberValue >= 0 && numberValue == trunc(numberValue);
+        if (!isUnsignedInteger) {
+            NSString *reason = [NSString stringWithFormat:@"Invalid value for key %@. Expected unsigned integer.", key];
+            [self failWithErrorCode:TWTJSONSchemaParserErrorCodeInvalidValue reason:reason object:number];
+        }
+    }
+
+    [self popPathComponent];
+    return number;
+}
+
+
+// Keywords: definitions
+- (NSDictionary *)parseDictionaryFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSDictionary *dictionary = schema[key];
+    [self checkObject:dictionary isOfClass:[NSDictionary class] allowNil:YES];
+    [self popPathComponent];
+    return dictionary;
+}
+
+
+// Keywords: enum
+- (NSArray *)parseArrayFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSArray *array = schema[key];
+    [self checkObjectIsArrayWithAtLeastOneItem:array allowNil:YES];
+    [self popPathComponent];
+    return array;
+}
+
+
+// Keyword: required
+// Example: { "required" : [ "name" ] }
+- (NSArray *)parseArrayOfUnqiueStringsFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSArray *array = schema[key];
+    [self checkObjectIsArrayWithAtLeastOneItem:array allowNil:YES];
+
+    NSMutableSet *uniqueStrings = [[NSMutableSet alloc] init];
+
+    [array enumerateObjectsUsingBlock:^(NSString *string, NSUInteger index, BOOL *stop) {
+        [self pushPathComponent:@(index)];
+        [self checkObject:string isOfClass:[NSString class] allowNil:NO];
+        if ([uniqueStrings containsObject:string]) {
+            NSString *reason = [NSString stringWithFormat:@"Invalid value for key %@. Expected array of unique strings.", key];
+            [self failWithErrorCode:TWTJSONSchemaParserErrorCodeRequiresUniqueness reason:reason object:string];
+        }
+
+        [uniqueStrings addObject:string];
+        [self popPathComponent];
+    }];
```

Comment:
I wonder if we can’t do this in a slightly nicer way. Ideally, if there were an error here, I’d want to see every string that appears more than once, not just the first one. This can sort of be accomplished using `NSCountedSet`:

```objc
NSArray *array = …;
NSCountedSet *countedSet = [NSCountedSet setWithArray:array];
NSSet *repeatedItems = [countedSet objectsPassingTest:^BOOL(id object, BOOL *stop) {
    return [countedSet countForObject:object] > 1;
}];

NSLog(@"Expected an array of unique items, but the following elements occur more than once: %@", [[repeatedItems allObjects] componentsJoinedByString:@", "]);
```

What do you think? The downside to this is that you lose the index in the error message. We could write some code that would do all of that, but it would be a little messy and is probably not worth the effort. What do you think?